ledger/simulation/tracer.go

```diff
@@ -163,9 +181,58 @@ func (tracer *evalTracer) saveApplyData(applyData transactions.ApplyData) {
 	applyDataOfCurrentTxn.EvalDelta = evalDelta
 }
 
+func (tracer *evalTracer) BeforeTxn(ep *logic.EvalParams, groupIndex int) {
+	if tracer.result.ExecTraceConfig > NoExecTrace {
+		// make transaction trace in following section
+		currentTxn := ep.TxnGroup[groupIndex]
+		traceType := OtherTransaction
+
+		if currentTxn.Txn.Type == protocol.ApplicationCallTx {
+			switch currentTxn.Txn.ApplicationCallTxnFields.OnCompletion {
+			case transactions.ClearStateOC:
+				traceType = AppCallClearStateTransaction
+			default:
+				traceType = AppCallApprovalTransaction
+			}
+		}
+		transactionTrace := TransactionTrace{TraceType: traceType}
+
+		var txnTraceStackElem *TransactionTrace
+
+		// The last question is, where should this transaction trace attach to:
+		// - if it is a top level transaction, then attach to TxnResult level
+		// - if it is an inner transaction, then refer to the stack for latest exec trace, and attach to inner array
+		if len(tracer.execTraceStack) == 0 {
+			tracer.result.TxnGroups[0].Txns[groupIndex].Trace = &transactionTrace
+			txnTraceStackElem = tracer.result.TxnGroups[0].Txns[groupIndex].Trace
+		} else {
+			lastExecTrace := tracer.execTraceStack[len(tracer.execTraceStack)-1]
+			lastExecTrace.InnerTraces = append(lastExecTrace.InnerTraces, transactionTrace)
+			txnTraceStackElem = &lastExecTrace.InnerTraces[len(lastExecTrace.InnerTraces)-1]
+
+			innerIndex := uint64(len(lastExecTrace.InnerTraces)) - 1
+			stepIndex := uint64(len(lastExecTrace.Trace)) - 1
+
+			lastExecTrace.StepToInnerMap = append(lastExecTrace.StepToInnerMap, TraceStepInnerIndexPair{
```

Comment:
Is `x.StepToInnerMap` saying that the `x.TraceStep` index of the `x.Trace` slice (which is `[]OpcodeTraceUnit`) spawns an inner transaction whose trace is the `x.InnerIndex` index of the `x.InnerTraces` slice?

Reply:
correct.

Reply:
This map between pc to inner index get moved into `ProgramTrace` object, so this thread resolving.

We can open new thread for designs in `OpcodeUnit` and `ProgramTrace`.